x-pack/plugins/security_solution/public/resolver/data_access_layer/mocks/pausify_mock.ts

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import {
+  ResolverRelatedEvents,
+  ResolverTree,
+  ResolverEntityIndex,
+} from '../../../../common/endpoint/types';
+import { DataAccessLayer } from '../../types';
+
+type PausableRequests = 'relatedEvents' | 'resolverTree' | 'entities';
+
+interface Metadata<T> {
+  /**
+   * The `_id` of the document being analyzed.
+   */
+  databaseDocumentID: string;
+  /**
+   * A record of entityIDs to be used in tests assertions.
+   */
+  entityIDs: T;
+}
+
+/**
+ * A simple mock dataAccessLayer that allows you to manually pause and resume a request.
+ */
+export function pausifyMock<T>({
+  metadata,
+  dataAccessLayer,
+}: {
+  dataAccessLayer: DataAccessLayer;
+  metadata: Metadata<T>;
+}): {
+  dataAccessLayer: DataAccessLayer;
+  metadata: Metadata<T>;
+  pause: (pausableRequests: PausableRequests) => void;
+  resume: (pausableRequests: PausableRequests) => void;
+} {
+  let relatedEventsPromise = Promise.resolve();
+  let resolverTreePromise = Promise.resolve();
+  let entitiesPromise = Promise.resolve();
+
+  let relatedEventsResolver: () => void;
+  let resolverTreeResolver: () => void;
+  let entitiesResolver: () => void;
+
+  return {
+    metadata,
+    pause: (pausableRequests: PausableRequests | PausableRequests[]) => {
+      const pauseRelatedEventsRequest = pausableRequests.includes('relatedEvents');
+      const pauseResolverTreeRequest = pausableRequests.includes('resolverTree');
+      const pauseEntitiesRequest = pausableRequests.includes('entities');
+
+      if (pauseRelatedEventsRequest) {
```

Comment:
I think this might not work if the code was called like this:
```
mock.pause(['resolverTree'])
mock.resolverTree().then(() => console.log('this will never run I think'))
mock.pause(['resolverTree'])
mock.resume(['resolverTree'])
```

Maybe you need to only create the promise if it hasn't already been created, and then set `relatedEventsPromise = null` when `resume` is called

Reply:
Updated per our convo, thanks